### ISSUE TITLE
No bidirectional inference for set()

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#3573]
### Summary

Ty recently added precise inference for unannotated empty lists, but not sets:

```python
from typing import reveal_type


def make_set() -> set[str]:
    result = reveal_type(set()) # set[Unknown]
    result.add(1)
    return result # accepted by ty (false negative)
```

**Comment by charliermarsh** [comment]
(Relevant here, I think, is that it's a function call and not a literal.)

**Comment by carljm** [comment]
Yeah I think to extend the current mechanism used for collection literals context, we'd have to recognize calls to `set()` in semantic indexing. Which is easy enough for the common case, but if we cared about handling aliased or shadowed `set`, that would be harder. But since this is an "improved precision" issue, it could be OK not to catch all those cases perfectly.

**Comment by jelle-openai** [comment]
The original example here actually still passes ty, but ty correctly catches the issue if you remove the `reveal_type()` call.

**Comment by carljm** [comment]
Yeah, I think that's a known limitation of how we need to recognize these constructors pre-type-checking in order to collect their later uses. It would definitely be possible to extend this so we can handle a constructor nested within another call, it's just a question of how far we extend it and what shapes we choose to recognize. Might be worth at least handling the `reveal_type` case.
